File: 1381-design-a-stack-with-increment-operation/1381-design-a-stack-with-increment-operation.java
class CustomStack {

    private List<Integer> stack;
    private int maxSize;

    public CustomStack(int maxSize) {
        // Initialize the stack as a LinkedList for efficient add/remove operations
        stack = new LinkedList<>();
        this.maxSize = maxSize;
    }

    public void push(int x) {
        // Add the element to the top of the stack if it hasn't reached maxSize
        if (stack.size() < maxSize) {
            stack.addLast(x);
        }
    }

    public int pop() {
        // Return -1 if the stack is empty, otherwise remove and return the top element
        if (stack.isEmpty()) return -1;
        return stack.removeLast();
    }

    public void increment(int k, int val) {
        // Create an iterator to traverse the stack from bottom to top
        ListIterator<Integer> iterator = stack.listIterator();

        // Increment the bottom k elements (or all elements if k > stack size)
        while (iterator.hasNext() && k > 0) {
            int current = iterator.next();
            iterator.set(current + val);
            k--;
        }
    }
}

File: 1381-design-a-stack-with-increment-operation/1381-design-a-stack-with-increment-operation.cpp
class CustomStack {
private:
    list<int> stack;
    int maxSize;

public:
    CustomStack(int maxSize) { this->maxSize = maxSize; }

    void push(int x) {
        // Add the element to the top of the stack if it hasn't reached maxSize
        if (stack.size() < maxSize) {
            stack.push_back(x);
        }
    }

    int pop() {
        // Return -1 if the stack is empty, otherwise remove and return the top
        // element
        if (stack.empty()) return -1;
        int topElement = stack.back();
        stack.pop_back();
        return topElement;
    }

    void increment(int k, int val) {
        // Increment the bottom k elements (or all elements if k > stack size)
        auto it = stack.begin();
        for (int i = 0; i < k && it != stack.end(); i++, it++) {
            *it += val;
        }
    }
};